pkg/types/validation/installconfig.go

```diff
@@ -330,8 +330,12 @@ func validateNetworking(n *types.Networking, singleNodeOpenShift bool, fldPath *
 		allErrs = append(allErrs, field.Required(fldPath.Child("networkType"), "network provider type required"))
 	}
 
-	if singleNodeOpenShift && n.NetworkType == string(operv1.NetworkTypeOpenShiftSDN) {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("networkType"), n.NetworkType, "networkType OpenShiftSDN is currently not supported on Single Node OpenShift"))
+	if n.NetworkType == string(operv1.NetworkTypeOpenShiftSDN) {
+		if singleNodeOpenShift {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("networkType"), n.NetworkType, "networkType OpenShiftSDN is currently not supported on Single Node OpenShift"))
+		} else {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("networkType"), n.NetworkType, "networkType OpenShiftSDN is deprecated, please use OVNKubernetes"))
```

Comment:
/hold

We're probably going to need some way to let customers install openshift-sdn with a support exception. Maybe by setting an environment variable (like with `OPENSHIFT_INSTALL_EXPERIMENTAL_DUAL_STACK`). Waiting to get some clarifications from PM...

Reply:
@danwinship The best way to allow that exception with no baggage in the installer is to tell users to run `create manifests` and then edit the network manifest. The installer will not validate user-supplied manifests. That is an extra step for users, so I'm not sure if it satisfactory, but it would work with this PR.

Reply:
Hm... ok, so there are 3 phases to this. (Sorry, this is still getting figured out on our side):
1. In 4.15, most customers should be blocked from creating openshift-sdn clusters, but (a) there needs to be some back door for customers with support exceptions, and (b) there needs to be a back door for the various `sdn` jobs in CI.
2. In 4.16, _all_ customers should be blocked from creating openshift-sdn clusters, and the back door that worked in 4.15 should not work any more. But CI still needs to be able to create openshift-sdn clusters, because we will still support openshift-sdn in 4.16 if you upgraded from 4.15.
3. In 4.17 openshift-sdn will be totally gone and nobody needs to be able to install new openshift-sdn clusters any more.

The "create manifests and then edit them" idea sounds like a good idea for the CI back door, although I'm not sure it's actually possible with our existing CI setup. (It seems like you can have a step that provides files to be copied into `manifests/`, but you can't have a step that runs arbitrary code in between "`create manifests`" and "`create cluster`"). (Though it's possible that every sdn-based CI job ends up with an identical `manifests/cluster-network-02-config.yml` anyway and so we could just unconditionally overwrite it?)

But that doesn't work as well for the customer back door, because of the constraint that the 4.15 solution is supposed to _not_ work in 4.16.

Reply:
> * In 4.15, most customers should be blocked from creating openshift-sdn clusters, but (a) there needs to be some back door for customers with support exceptions, and (b) there needs to be a back door for the various `sdn` jobs in CI.

I agree with Patrick that generating the manifests and editing them is the right way here (and conveniently already exists).

> * In 4.16, _all_ customers should be blocked from creating openshift-sdn clusters, and the back door that worked for support exceptions in 4.15 should not work any more.

There are _so_ many things you _can_ do (i.e. are not prevented by a validation) that nevertheless aren't supported that I'm not sure this one is special. In general we do a lot of validation on the install-config but little to none on any manifests that users edit.

If you are doing a thing because you have a support exception for it and that support exception expires, then if you keep doing it you will no longer be supported. This shouldn't come as a surprise to anybody that has an explicit support exception.

> But CI still needs to be able to create openshift-sdn clusters, because we will still support openshift-sdn in 4.16 if you upgraded from 4.15.

Shouldn't CI be installing 4.15 clusters and upgrading them to 4.16 to test openshift-sdn if that's the only scenario we care about?

Reply:
OK, actually, here's the plan (OK'ed by PM):
1. In 4.15, we drop `OpenShiftSDN` support from `install-config.yaml` as in this PR, and we will tell users with support exceptions that they can work around it by creating a manifest for `network.config.openshift.io`.
2. In 4.16, we will make CNO refuse to copy the `network.config.openshift.io` config over to a new `network.operator.openshift.io` config if the network type is `OpenShiftSDN`, which means the manifest hack we documented for 4.15 will no longer work. (It _will_ still work if you create manifests for _both_ the `network.config.openshift.io` and the `network.operator.openshift.io` objects rather than having CNO construct the operator config from the config config, and this is what CI will do, but this won't ever have been documented as something customers should do.)

Reply:
(github hadn't updated so I didn't see your comment when I replied before.)

> There are _so_ many things you _can_ do (i.e. are not prevented by a validation) that nevertheless aren't supported that I'm not sure this one is special.

The thing that's special about this one is the fact that we will have actually _told_ people to do it, so it feels more like "I'm using a special case" than "I'm cheating". But anyway, we can block that special case from outside of the installer like I said above, so we're good here.

> Shouldn't CI be installing 4.15 clusters and upgrading them to 4.16 to test openshift-sdn if that's the only scenario we care about?

:man_shrugging: OCP in general assumes that that it can treat those two things as equivalent (as long as the 4.x-to-4.x+1 upgrade jobs are passing), and it's so much faster ( / cheaper) to _not_ test it that way.